### PR TITLE
User stopped typing behavior

### PR DESCRIPTION
**Pull request by jBijsterboschNL**
### Description of Change ###

Implemented UserStoppedTypingBehavior. This behavior waits for the user to stop typing and then executes a Command. The waiting threshold is adjustable, as well as there is an option to dismiss the keyboard after the user has stopped typing.

### Bugs Fixed ###

- Related to issue #223

### API Changes ###

None

### Behavioral Changes ###

None

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [ ] Changes adhere to coding standard
- [ ] Updated documentation

### CODE REVIEW

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:21]: no need to write ```defaultValue: 1000``` just ```1000``` )

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:28]: same here

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:22]: I think it makes sense to clarify what this means.
StoppedTypingTimeThreshold or StoppedTypingMillisecondsThreshold

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:29]: @jfversluis @sthewissen @Char0394 
Perhaps it's only my personal preference, but I prefer to use "Is", "Are", "Should" as infix for all boolean properties.

e.g. ShouldDismissKeyboardAutomatically or IsKeyboardAutoDismissing

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:55]: OnInputViewTextChanged

But I propose you to override existing OnViewPropertyChanged. There you will be able to cast "View" and retrieve necessary property

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:69]: Just for better readability during the review please update the formatting in this file :)

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:63]: I think we needn't have to pass tokenSource.Token here. We can get token inside the PerformTextChanged's body

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:71]: How about re-writing this method to 
```
await Task.Delay(StoppedTypingThreshold, cancellationToken.Token)
.ContinueWith(task => { 
    if (task.Exception != null) return;
    /// Proceed
});
```
Or even try-catch all method (but anyway I suggest to pass token as a parameter to Task.Delay method).

**Review by AndreiMisiukevich — CHANGES_REQUESTED** [review]
Hi, thank you for this awesome contribution. I appreciate any desire to make XF World better and more convenient. Especially when the contribution is covered by unit tests and has a sample 👍

File: XamarinCommunityToolkitSample.sln
Comment (jBijsterboschNL) [reply XamarinCommunityToolkitSample.sln:9]: Not sure where these changes came from. After merging in latest version from main, this got changed as well. When I undo this, and open the solution again, Visual Studio for Windows automatically does the same changes again. Not sure if it does any harm? Also not sure how I could prevent this?

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (jBijsterboschNL) [reply XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:55]: Hi @AndreiMisiukevich, but why would you want to override the OnViewPropertyChanged for this, as the InputView already has a specific event (TextChanged) for this? Otherwise the OnViewPropertyChanged code will be called for all property changes, not only for the Text property to change, and we need to add a check if the Text property has changed. I don't see the benefits of it and why that would be a better solution. The naming change has been made (once I commit my changes).

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (jBijsterboschNL) [reply XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:63]: Hi @AndreiMisiukevich, this is definitely needed like so. As otherwise the token cancellation won't work properly no more. After applying this change requested by you, one of the unit tests failed (thanks for writing unit tests, yay!).

**Comment by jBijsterboschNL** [comment]
Hi @AndreiMisiukevich, thank you for your quick code review action. I've applied changes that you requested and replied to some change request to discuss further.

File: XamarinCommunityToolkit.UnitTests/Behaviors/UserStoppedTypingBehavior_Tests.cs
Comment (pictos) [inline XamarinCommunityToolkit.UnitTests/Behaviors/UserStoppedTypingBehavior_Tests.cs:14]: Any reason to don't use `async Task` here and in the others methods?

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (pictos) [reply XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:63]: Also, instead of using `sender as Entry` can you use `(Entry)sender`. Because if the View isn't an `Entry` we can face a null exception at `entry.Unfocu() //line 74`. So I think it's better to fail here because of the sync/async behavior. Do you agree?

Comment (jBijsterboschNL) [reply XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:63]: Hi @pictos, actually the type check has already been done in the OnAttached() method. So that is not possible. But I did see that I was casting to an Entry instead of an InputView, so will change that here.

**Comment by jBijsterboschNL** [comment]
@AndreiMisiukevich and @pictos, I've applied all requested changed from your code reviews. Could you guys re-check and approve this PR once you guys agree with the code changes that have been made? Cheers.

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs]: in that case, you won't have to take care of unsubscribing from your event.

```csharp
protected override void OnViewPropertyChanged(object sender, PropertyChangedEventArgs e) {
    if (e.PropertyName == InputView.TextProperty.PropertyName) {
        var text = ((InputView)View).Text;
    }
}
```

but uh

File: XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Behaviors/UserStoppedTypingBehavior.shared.cs:55]: please rename it to OnInputTextChanged

File: XamarinCommunityToolkitSample/ViewModels/Behaviors/UserStoppedTypingBehaviorViewModel.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkitSample/ViewModels/Behaviors/UserStoppedTypingBehaviorViewModel.cs:25]: how about

```=> SearchCommand = new Command<string>(PerformSearch);```
For this ctor and for PerformSearch as well

**Review by sthewissen — APPROVED** [review]
Awesome! Let's merge this!

**Review by AndreiMisiukevich — APPROVED** [review]
LGTM
Good work with comments :) 👍

**Comment by jfversluis** [comment]
I've updated the PR template to automatically also resolve the issue when the PR is merged. That is typically what we want.

Great work @jBijsterboschNL !

**Pull request merged**
